text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+should have a clear meaning, and `impl(self)` and `mut(self)` are not clear. While they should be
+allowed for consistency with `pub(self)`, an alternative was needed with the same behavior.
+`impl(mod)` and `mut(mod)` should have a sufficiently clear meaning: implementing the trait and
+mutating the field are restricted to the current module. `impl(in mod)` and `mut(in mod)` are
+accepted as well. The behavior of `impl(mod)` is identical to `impl(self)`; likewise for `mut(mod)`
+and `mut(self)`.
+
+For consistency with the new restriction syntax, `pub(mod)` and `pub(in mod)` are also allowed.
+
+# Reference-level explanation
+
+## Syntax
+
+Using the syntax from [the reference for `struct`s][struct syntax], the change needed to support
+`mut` restrictions is quite small.
+
+[struct syntax]: https://doc.rust-lang.org/stable/reference/items/structs.html
+
+```diff
+StructField :
+   OuterAttribute*
+   Visibility?
++  MutRestriction?
+   IDENTIFIER : Type
+
+TupleField :
+   OuterAttribute*
+   Visibility?
++  MutRestriction?
+   Type
+
++MutRestriction :
++     mut
++   | mut ( crate )
++   | mut ( self )
++   | mut ( mod )
++   | mut ( super )
++   | mut ( in SimplePath )
++   | mut ( in mod )
+```
+
+Trait definitions need a similar change to the [syntax for `trait`s][trait syntax] to accommodate
+`impl` restrictions.
+
+[trait syntax]: https://doc.rust-lang.org/stable/reference/items/traits.html
+
+```diff
+Trait :
+   unsafe?
++  ImplRestriction?
+   trait IDENTIFIER
+   GenericParams? ( : TypeParamBounds? )? WhereClause? {
+     InnerAttribute*
+     AssociatedItem*
+   }
+
++ImplRestriction :
++     impl
++   | impl ( crate )
++   | impl ( self )
++   | impl ( mod )
++   | impl ( super )
++   | impl ( in SimplePath )
++   | impl ( in mod )
+```
+
+Essentially, `mut` and `impl` have the same syntax as `pub`, just with a different keyword.
+
+Finally, a minor change is needed to the syntax for visibility to permit `pub(mod)`.
+
+```diff
+Visibility :
+     pub
+   | pub ( crate )
+   | pub ( self )
++  | pub ( mod )
+   | pub ( super )
+   | pub ( in SimplePath )
++  | pub ( in mod )
+```
+
+## Behavior
+
+The current behavior of `pub` is that `pub` makes something visible within the declared scope. If no
+scope is declared (such that it is just `pub`), then the item is visible everywhere. This behavior
+is preserved for `impl` and `mut`. When a restriction is used, the behavior is allowed only within
+the declared scope. When the keyword is used alone, the behavior is allowed everywhere. While in
+most cases the default visibility is private, `pub` is default in some cases, namely `enum`
+variants, `enum` fields, and `trait` items. `impl` and `mut` will have a consistent default: when
+omitted entirely, the scope is inherited from `pub`. This is both what is most convenient and is
+what is required for backwards compatibility with existing code.
```

Comment:
This seems really confusing, in addition to being redundant. In particular, `pub impl Trait { ... }` seems really easy to confuse with `impl Trait for ...`.

This would become even *more* confusing if, in the future, we support trait implementations with limited visibility (`pub(crate) impl Trait for Type { ... }`).

I think we shouldn't support bare `impl` or `mut` at all. Is there some specific use case that motivates them, where they wouldn't be redundant?